src/cata_ui.cpp

```diff
@@ -0,0 +1,1038 @@
+{
+}
+
+void ui_vertical_list::draw()
+{
+    auto win = get_win();
+    if( win == nullptr ) {
+        return;
+    }
+
+    unsigned int start_line = window_scroll;
+    unsigned int end_line = start_line + get_rect().size_y;
+
+    unsigned int available_space = get_rect().size_x - 2; // 2 for scroll bar and spacer
+
+    for( unsigned int line = start_line; line < end_line && line < text.size(); line++ ) {
+        auto txt = text[line];
+        if( txt.size() > available_space ) {
+            txt = txt.substr( 0, available_space );
+        }
+        if( scroll == line ) {
+            mvwprintz( win, get_ay() + line - start_line, get_ax() + 2, hilite(text_color), txt.c_str() );
+        } else {
+            mvwprintz( win, get_ay() + line - start_line, get_ax() + 2, text_color, txt.c_str() );
+        }
+    }
+
+    draw_scrollbar( win, scroll, get_rect().size_y, text.size(), get_ay(), get_ax(), bar_color, false );
+}
+
+void ui_vertical_list::set_text( std::vector<std::string> text )
+{
+    this->text = text;
+}
+
+void ui_vertical_list::scroll_up()
+{
+    scroll = ( scroll == 0 ? text.size() - 1 : scroll - 1 );
+
+    if( scroll == text.size() - 1 ) {
+        window_scroll = scroll - get_rect().size_y + 1;
+    } else if( scroll < window_scroll ) {
+        window_scroll--;
+    }
+}
+
+void ui_vertical_list::scroll_down()
+{
+    scroll = ( scroll == text.size() - 1 ? 0 : scroll + 1 );
+
+    if( scroll > get_rect().size_y + window_scroll - 1 ) {
+        window_scroll++;
+    } else if( scroll == 0 ) {
+        window_scroll = 0;
+    }
+}
+
+const std::string &ui_vertical_list::current() const
+{
+    return text[scroll];
+}
+
+ui_horizontal_list::ui_horizontal_list( int x, int y, ui_anchor anchor ) : ui_element( 0, 1, x, y, anchor )
+{
+}
+
+void ui_horizontal_list::draw()
+{
+    auto win = get_win();
+    if( win == nullptr ) {
+        return;
+    }
+
+    unsigned int x_offset = 1 + get_ax();
+
+    for( unsigned int i = 0; i < text.size(); i++ ) {
+        x_offset += draw_subtab( win, get_ax() + x_offset, get_ay(), text[i], i == scroll ) + 2;
+    }
+}
+
+void ui_horizontal_list::set_text( std::vector<std::string> text )
+{
+    this->text = text;
+    size_t text_length = 1;
+
+    for( const auto &str : text ) {
+        text_length = utf8_width( str ) + 1;
+    }
+
+    set_rect( ui_rect( text_length, get_rect().size_y, get_rect().x, get_rect().y ) );
+}
+
+void ui_horizontal_list::scroll_left()
+{
+    scroll = ( scroll == 0 ? text.size() - 1 : scroll - 1 );
+}
+
+void ui_horizontal_list::scroll_right()
+{
+    scroll = ( scroll == text.size() - 1 ? 0 : scroll + 1 );
+}
+
+const std::string &ui_horizontal_list::current() const
+{
+    return text[scroll];
+}
+
+/////////////////////////////////////////////////////
+//THE FOLLOWING IS TEMPORARY/////////////////////////
+/////////////////////////////////////////////////////
+void label_test()
+{
+    auto lable1 = new ui_label("some", 0, 0, top_left);
```

Comment:
I saw your comment about using uique_ptr to the last guy that made a ui framework. I don't fully understand them, so I used what I know.

Back then, you said something about the pointer not being able to be shared? Does that mean that there can only be 1 pointer to the pointed to object?

I want to have it so that you can add ui_elements to a window, but still change them with the pointer you have, so you can do updates to it's values after adding. Would that still be possible?